### PR TITLE
 Add triplet loss for metric lerning.

### PR DESCRIPTION
**Pull request by wanghaoshuang**
前向的计算过程：https://wanghaoshuang.github.io/2018/07/Dense-Triplet_Loss/

### CODE REVIEW

**Comment by wanghaoshuang** [comment]
@BigFishMaster  你那里有`dense_triplet_loss`比较专业的英文文档么？

**Comment by BigFishMaster** [comment]
@wanghaoshuang 我稍后看看哈

File: paddle/fluid/operators/dense_triplet_loss_op.h [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+namespace operators {
+
+using Tensor = framework::Tensor;
+template <typename T, int MajorType = Eigen::RowMajor,
+          typename IndexType = Eigen::DenseIndex>
+using EigenMatrix = framework::EigenMatrix<T, MajorType, IndexType>;
+
+template <typename T, int MajorType = Eigen::RowMajor,
+          typename IndexType = Eigen::DenseIndex>
+using EigenVector = framework::EigenVector<T, MajorType, IndexType>;
```
Comment (qingqing01) [inline paddle/fluid/operators/dense_triplet_loss_op.h]: `EigenVector` is not used in following code.

File: paddle/fluid/operators/dense_triplet_loss_op.h [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    auto d_logits_t = EigenMatrix<T>::From(*d_logits);
+    auto loss_t = EigenMatrix<T>::From(*loss);
+    auto distances_t = EigenMatrix<T>::From(distances);
+    auto d_distances_t = EigenMatrix<T>::From(d_distances);
+    auto tmp_t = EigenMatrix<T>::From(tmp);
+
+    auto blas = math::GetBlas<DeviceContext, T>(context);
+    auto x_mat = math::CreateMatrixDescriptor(x_dims, 0, false);
+    auto x_mat_trans = math::CreateMatrixDescriptor(x_dims, 0, true);
+    blas.MatMul(*logits, x_mat, *logits, x_mat_trans, T(1), &distances, T(0));
```
Comment (qingqing01) [inline paddle/fluid/operators/dense_triplet_loss_op.h]: Since the input is 2D tensor, there is no need to use `math::CreateMatrixDescriptor`,  refer the MatMul inferface:
 https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/math/blas_impl.h#L176

File: paddle/fluid/operators/dense_triplet_loss_op.h [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+template <typename T, int MajorType = Eigen::RowMajor,
+          typename IndexType = Eigen::DenseIndex>
+using EigenMatrix = framework::EigenMatrix<T, MajorType, IndexType>;
+
+template <typename T, int MajorType = Eigen::RowMajor,
+          typename IndexType = Eigen::DenseIndex>
+using EigenVector = framework::EigenVector<T, MajorType, IndexType>;
+
+using DIM1 = Eigen::array<int, 1>;
+using DIM2 = Eigen::array<int, 2>;
```
Comment (qingqing01) [inline paddle/fluid/operators/dense_triplet_loss_op.h]: DIM1 -> Array1 ?
DIM2 -> Array2?

File: paddle/fluid/operators/dense_triplet_loss_op.cc [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    AddComment(R"DOC(
+
+The Input(logits) is the embeddings of input samples. And the triplet loss is 
+defined over triplets of embeddings. A triplet contains three samples as below:
+  - an anchor sample
+  - a positive sample with the same class as the anchor
+  - a negative sample with a different class
+We define the three samples as $a$, $p$, $n$. Then the loss of
+the triplet (a, p, n) is:
+  $$L = max(d(a, p) - d(a, n) + margin, 0)$$
```
Comment (qingqing01) [inline paddle/fluid/operators/dense_triplet_loss_op.cc:65]: 前后留空格。

File: paddle/fluid/operators/dense_triplet_loss_op.cc [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    AddOutput("Loss",
+              "(Tensor, default: Tensor<float>), A 2-D tensor. The triplet "
+              "loss with shape [N x 1].");
+    AddOutput("LogitsGrad",
+              "(Tensor, default: Tensor<float>), A temporary "
+              "output Tensor to store the gradients of triplet loss, which is "
+              "computed with loss together in one call. It is a 2-D Tensor of "
+              "the shape [N, feature_len].")
+        .AsIntermediate();
+    AddAttr<float>("margin", "(float), The min margin between two sample.");
```
Comment (qingqing01) [inline paddle/fluid/operators/dense_triplet_loss_op.cc:54]: min -> minimum

File: paddle/fluid/operators/dense_triplet_loss_op.cc [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+The Input(logits) is the embeddings of input samples. And the triplet loss is 
+defined over triplets of embeddings. A triplet contains three samples as below:
+  - an anchor sample
+  - a positive sample with the same class as the anchor
+  - a negative sample with a different class
+We define the three samples as $a$, $p$, $n$. Then the loss of
+the triplet (a, p, n) is:
+  $$L = max(d(a, p) - d(a, n) + margin, 0)$$
+In which, $d(a, p)$ means the distance between $a$ and $p$. The negative should
```
Comment (qingqing01) [inline paddle/fluid/operators/dense_triplet_loss_op.cc:66]: distance -> L2 distance ?

File: python/paddle/fluid/layers/nn.py [hunk 7]
```diff
@@ -5184,3 +5185,52 @@ def crop(x, shape=None, offsets=None, name=None):
         outputs={'Out': out},
         attrs=None if len(attrs) == 0 else attrs)
     return out
+
+
+def dense_triplet_loss(input, label, margin=0.01):
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py:5190]: Add name:

```
dense_triplet_loss(input, label, margin=0.01, name=None)
```

File: paddle/fluid/operators/dense_triplet_loss_op.cc [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+class DenseTripletLossOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("Logits",
+             "(Tensor, default: Tensor<float>), A 2-D tensor with shape [N x "
+             "K]. N is the total number of samples, "
+             "and K is the feature length in each sample.");
+    AddInput("Label",
+             "(Tensor) The ground truth which is a 2-D tensor.  "
+             "Label is a Tensor<int64> with shape [N x 1]. ");
```
Comment (qingqing01) [inline paddle/fluid/operators/dense_triplet_loss_op.cc:44]: 这里要求Label是按index排过序的吗？

File: python/paddle/fluid/tests/unittests/test_triplet_loss_op.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    return loss.reshape([batch_size, 1])
+
+
+class TestTripletOp(OpTest):
+    def setUp(self):
+        self.op_type = "dense_triplet_loss"
+        self.batch_size = 9
+        self.feature_len = 3
+        self.class_num = 4
+        self.eps = 0.01
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/test_triplet_loss_op.py:54]: self.eps -> self.margin ?

File: python/paddle/fluid/tests/unittests/test_triplet_loss_op.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+    def test_check_output(self):
+        self.check_output()
+
+    def test_check_grad(self):
+        self.outputs["LogitsGrad"] = self.logits_grad
+        self.check_grad(
+            ["Logits"],
+            "Loss",
+            max_relative_error=0.05,
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/test_triplet_loss_op.py:75]: forward里有relu，这里的gradient check有随机错误吗？
